test/extended/util/cmonitor.go

```diff
@@ -0,0 +1,325 @@
+	if err != nil {
+		return nil, fmt.Errorf("Failed to obtain a lease: %v", err)
+	}
+	leaseID := latest.Node.Value
+	for i := 0; i < len(m.alive); {
+		ctrls := m.alive[i]
+		if ctrls.Exited() {
+			m.markDead(ctrls)
+			continue
+		}
+		lid, err := ctrls.GetLeaseID(false)
+		if err != nil {
+			fmt.Fprintf(g.GinkgoWriter, "Failed to get a leaseID of %s: %v\n", ctrls.String(), err)
+		} else if leaseID == lid {
+			return ctrls, nil
+		}
+		i++
+	}
+	return nil, nil
+}
+
+func (m *CMonitor) WaitForActive(timeout time.Duration) (*Controllers, time.Duration, error) {
+	var latestResp *etcdclient.Response
+	start := time.Now()
+	findInstance := func(leaseID string) *Controllers {
+		var lid string
+		for i := 0; i < len(m.alive); {
+			if m.alive[i].Exited() {
+				m.markDead(m.alive[i])
+				continue
+			}
+			ec := make(chan error)
+			go func() {
+				var err error
+				lid, err = m.alive[i].GetLeaseID(true)
+				ec <- err
+			}()
+			select {
+			case err := <-ec:
+				if err != nil {
+					fmt.Fprintf(g.GinkgoWriter, "Failed to get a leaseID of %s: %v\n", m.alive[i], err)
+				} else if leaseID == lid {
+					return m.alive[i]
+				}
+			case <-time.After(start.Add(timeout).Sub(time.Now())):
+				return nil
+			}
+			i++
+		}
+		leases := []string{}
+		for i := 0; i < len(m.alive); i++ {
+			if lease, err := m.alive[i].GetLeaseID(false); err != nil {
+				leases = append(leases, lease)
+			}
+		}
+		fmt.Fprintf(g.GinkgoWriter, "Current LeaseID %q does not belong to any controllers instance running: %s\n", leaseID, strings.Join(leases, ", "))
+		return nil
+	}
+
+	latestResp, err := m.EtcdClient.Get(EtcdLeasePath, false, false)
+	if err != nil && !strings.Contains(strings.ToLower(err.Error()), "key not found") {
+		return nil, time.Now().Sub(start), fmt.Errorf("Failed to obtain a lease: %v", err)
+	}
+	if err == nil {
+		if ctrls := findInstance(latestResp.Node.Value); ctrls != nil {
+			return ctrls, time.Now().Sub(start), nil
+		}
+	}
+	now := time.Now()
+	if start.Add(timeout).Before(now) {
+		return nil, now.Sub(start), fmt.Errorf("timeout (%s) occured while waiting for a lease to be set", timeout.String())
+	}
+
+Loop:
+	for {
+		// TODO: is there a need for a stop channel?
+		stopChan := make(chan bool)
+		responseChan := make(chan *etcdclient.Response)
+		errChan := make(chan error)
+		index := uint64(0)
+		if latestResp != nil {
+			index = latestResp.Node.ModifiedIndex + 1
+		}
+		go func() {
+			response, err := m.EtcdClient.Watch(EtcdLeasePath, index, false, nil, stopChan)
+			if err != nil {
+				errChan <- err
+			}
+			responseChan <- response
+		}()
+		select {
+		case err := <-errChan:
+			return nil, time.Now().Sub(start), err
+		case resp := <-responseChan:
+			if ctrls := findInstance(resp.Node.Value); ctrls != nil {
+				return ctrls, time.Now().Sub(start), nil
+			}
+			latestResp = resp
+		case <-time.After(start.Add(timeout).Sub(time.Now())):
+			stopChan <- true
```

Comment:
stopchans are supposed to broadcast so you should close this instead. I don't see it used anywhere though.